src/macaron/slsa_analyzer/package_registry/pypi_registry.py

```diff
@@ -0,0 +1,198 @@
+# Copyright (c) 2023 - 2024, Oracle and/or its affiliates. All rights reserved.
+# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/.
+
+"""The module provides abstractions for the pypi package registry."""
+
+import logging
+import os
+import re
+from datetime import datetime
+from typing import Any
+
+from bs4 import BeautifulSoup
+
+from macaron.slsa_analyzer.package_registry.package_registry import PackageRegistry
+from macaron.util import send_get_http, send_get_http_raw
+
+logger: logging.Logger = logging.getLogger(__name__)
+
+
+class PyPIRegistry(PackageRegistry):
+    """This class implements the npm package registry.
+
+    There is no complete and up-to-date API documentation for the npm registry and the endpoints
+    are discovered by manual inspection of links on https://www.npmjs.com.
+    """
+
+    def __init__(
+        self,
+        hostname: str | None = None,
+        attestation_endpoint: str | None = None,
+        request_timeout: int | None = None,
+        enabled: bool = True,
+    ) -> None:
+        """
+        Initialize the npm Registry instance.
+
+        Parameters
+        ----------
+        hostname : str | None
+            The hostname of the npm registry.
+        attestation_endpoint : str | None
+            The attestation REST API.
+        request_timeout : int | None
+            The timeout (in seconds) for requests made to the package registry.
+        enabled: bool
+            Shows whether making REST API calls to npm registry is enabled.
+        """
+        self.hostname = hostname or ""
+        self.attestation_endpoint = attestation_endpoint or ""
+        self.request_timeout = request_timeout or 10
+        self.enabled = enabled
+        super().__init__("pypi Registry")
+
+
+class PyPIApiClient:
+    """PyPI's official and customed API client."""
+
+    def __init__(self, package: str) -> None:
+        self.headers = {"Host": "pypi.org", "Accept": "application/json"}
+        self.base_url = "https://pypi.org"
+        self.package = package
+        self.base_metadata: dict[str, Any] = self._get_base_metadata()
+
+    def _get_base_metadata(self) -> dict:
+        """Get the metadata through PyPI official API.
+
+        Returns
+        -------
+            dict | None: PyPI official metadata.
+        """
+        endpoint = os.path.join(self.base_url, "pypi", self.package, "json")
+        response: dict = send_get_http(endpoint, headers=self.headers)
+        return response
+
+    def get_releases(self) -> dict | None:
+        """Get all releases.
+
+        Returns
+        -------
+            dict | None: Version to metadata.
+        """
+        releases: dict | None = self.base_metadata.get("releases", None)
+        return releases
```

Comment:
Some of the methods here I can see that they are using the base metadata obtained from the method call `_get_base_metadata`. From this comment https://github.com/oracle/macaron/pull/750/files#r1632579056, and also it's usually introduce more complexity by adding states to a class instance (the state in this case is represented by different `self.base_metadata`, which changes continuously depending on the package we are analyzing). Therefore, I think we should:
- Remove the use of storing the result of `base_metadata` as a class attribute.
- Because a lot of information can be extracted from the `base_metadata` dictionary, it's not good to keep performing the API again and again, let's change these methods signature so that it accepts the `response` as a parameter
```python
def get_releases(self, response: dict) -> dict | None:
    releases: dict | None = response.get("releases", None)
    return releases
```
Then within the check, we will keep the value of `response` and use it again and again
```
base_metadata = pypi_registry.get_base_metadata(package_name)
releases = pypi_registry.get_releases(base_metadata)
project_links = pypi_registry.get_project_links(base_metadata)
```

Reply:
I additions, if you are working with a json like data structure (a dictionary that was loaded from a json content, i.e the API result), we have this utility function to extract values from it - https://github.com/oracle/macaron/blob/e214326697aee6e31f0c9e406b11344c8619581e/src/macaron/json_tools.py#L15.
For example, let's use this function for `get_releases`
```python
from macaron.json_tools import json_extract
...
def get_releases(self, base_metadata: dict) -> dict | None:
        """Get all releases.
        Returns
        -------
            dict | None: Version to metadata.
        """
        releases = json_extract(base_metadata, ["releases"], dict)
        return releases
```
We should consider changing other methods (e.g. `get_latest_version`, etc.) to use this `json_extract` method too).

Reply:
Fixed this in f900dfef1bae0413a4474f9e8e9101db3c625af1